File: rust/src/leetcode/Q49_group_anagrams.rs
/*
 * @lc app=leetcode id=49 lang=rust
 *
 * [49] Group Anagrams
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut map: HashMap<String, Vec<String>> = HashMap::new();

        for s in strs {
            let mut chars: Vec<char> = s.chars().collect();
            chars.sort();
            let key: String = chars.into_iter().collect();

            map.entry(key).or_default().push(s);
        }

        map.into_values().collect()
    }
}
// @lc code=end

File: python/Q49_group_anagrams.py
#
# @lc app=leetcode id=49 lang=python3
#
# [49] Group Anagrams
#


# @lc code=start
class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        result: dict[str, list[str]] = dict()

        for s in strs:
            key: str = "".join(sorted(s))

            if key not in result:
                result[key] = []

            result[key].append(s)

        return list(result.values())


# @lc code=end
